components/esp_dns/esp_dns.c

```diff
@@ -0,0 +1,318 @@
+        break;
+
+    default:
+        ESP_LOGW(TAG, "Unknown protocol during cleanup: %d", handle->config.protocol);
+        ret = -1;
+        break;
+    }
+
+    /* Mark as uninitialized */
+    handle->initialized = false;
+
+    /* Release and delete mutex */
+    xSemaphoreGive(handle->lock);
+    vSemaphoreDelete(handle->lock);
+
+    /* Take global mutex before modifying global handle */
+    if (g_dns_global_mutex != NULL && xSemaphoreTake(g_dns_global_mutex, portMAX_DELAY) == pdTRUE) {
+        /* Clear global handle if it matches this one */
+        if (g_dns_handle == handle) {
+            g_dns_handle = NULL;
+        }
+
+        xSemaphoreGive(g_dns_global_mutex);
+    }
+
+    /* Empty the handle */
+    memset(handle, 0, sizeof(handle));
+
+    ESP_LOGI(TAG, "DNS module cleaned up %s", (ret == 0) ? "successfully" : "with errors");
+    return ret;
+}
+
+/**
+ * Get the last error code from the DNS module.
+ *
+ * @return Last error code
+ */
+int dns_get_last_error(esp_dns_handle_t handle)
+{
+    if (handle == NULL) {
+        return DNS_ERR_INIT_FAILED;
+    }
+
+    return handle->last_error;
+}
+
+
+/* ========================= LWIP HOOK FUNCTIONS ========================= */
+
+#if defined(CONFIG_LWIP_HOOK_NETCONN_EXT_RESOLVE_CUSTOM)
+int lwip_hook_netconn_external_resolve(const char *name, ip_addr_t *addr, u8_t addrtype, err_t *err)
+{
+    if (g_dns_handle == NULL) {
+        ESP_LOGD(TAG, "ESP_DNS module not initialized, resolving through native DNS");
+        *err = ERR_OK;
+        return 0;
+    }
+
+    if (name == NULL || addr == NULL || err == NULL) {
+        if (err) {
+            *err = ERR_ARG;
+        }
+        return 1;
+    }
+
+    /* Check if name is already an IP address */
+    if (ipaddr_aton(name, addr)) {
+        *err = ERR_OK;
+        return 0;
+    }
+
+    /* Check if HTTPS_DNS_SERVER is in the dns cache */
+    if ((strcmp(name, g_dns_handle->config.dns_server) == 0) ||
+#if LWIP_HAVE_LOOPIF
+            /* Check if it's localhost */
+            (strcmp(name, "localhost") == 0) ||
+#endif
+            ipaddr_aton(name, addr)) { /* host name already in octet notation */
+        return 0;
+    }
+
+    u8_t rrtype;
+    if ((addrtype == NETCONN_DNS_IPV4) || (addrtype == NETCONN_DNS_IPV4_IPV6)) {
+        rrtype = DNS_RRTYPE_A;
+    } else if ((addrtype == NETCONN_DNS_IPV6) || (addrtype == NETCONN_DNS_IPV6_IPV4)) {
+        rrtype = DNS_RRTYPE_AAAA;
+    } else {
+        ESP_LOGE(TAG, "Invalid address type");
+        *err = ERR_VAL;
+        return 1;
+    }
+
+    /* Resolve based on configured transport type */
+    switch (g_dns_handle->config.protocol) {
+    case DNS_PROTOCOL_UDP:
+        /* Return zero as lwip dns can handle dns udp */
+        return 0;
+    case DNS_PROTOCOL_TCP:
+        *err = dns_resolve_tcp(g_dns_handle, name, addr, rrtype);
+        break;
+    case DNS_PROTOCOL_DOT:
+        *err = dns_resolve_dot(g_dns_handle, name, addr, rrtype);
+        break;
```

Comment:
Question, can we use the factory design pattern here? (e.g. if a user wants DNS over TCP, they don't need to pull mbedtls into the link dependency)

Reply:
Expose all the init function and remove the common init.

Reply:
Done.